### PR TITLE
feat: add sqlglot to parse sql dataflow

### LINKED ISSUE #3103: More comprehensive parsing of the SQL dataflow graph
### Description

We fail to parse Tables and Common Table Expression (CTE) in certain nested expressions or joins in SQL. 

We would like to be able to cover more edge-case. They can be found in:
* `tests/_ast/test_sql_visitor.py` > "@pytest.mark.xfail"

### Suggested solution

We could continue parsing ourselves using the token/ast that duckdb returns, or use a more comprehensive library like [sqlglot](https://github.com/tobymao/sqlglot)

### Alternative

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Light2Dark**
## 📝 Summary

Fixes #3103. Adds sqlglot as an optional dependency to handle complex sql refs parsing. Also supports DML statements (insert, update and delete).

## 🔍 Description of Changes

- set sqlglot version to lowest possible (likely >23.4, due to ibis lib)
- If we're keen on supporting dml's, then I could add more tests (I added some here alrdy)
- `def find_sql_def()` can also be modified to use sqlglot in the future, not included in this PR.
- small change: updated duckdb version to 1.1.0 to be synchronised across all installations (some were 1.0 and some were 1.1)

## 📋 Checklist

- [x] I have read the [contributor guidelines](https://github.com/marimo-team/marimo/blob/main/CONTRIBUTING.md).
- [ ] For large changes, or changes that affect the public API: this change was discussed or approved through an issue, on [Discord](https://marimo.io/discord?ref=pr), or the community [discussions](https://github.com/marimo-team/marimo/discussions) (Please provide a link if applicable).
- [x] I have added tests for the changes made.
- [x] I have run the code and verified that it works as expected.

## 📜 Reviewers

@akshayka OR @mscolnick

### CODE REVIEW

File: tests/_ast/test_sql_visitor.py [hunk 1]
```diff
@@ -593,7 +601,7 @@ def test_find_sql_refs_with_lateral_join() -> None:
         SELECT * FROM employees,
         LATERAL (SELECT * FROM departments WHERE departments.id = employees.dept_id) dept;
         """
-        assert find_sql_refs(sql) == ["employees", "departments"]
+        assert find_sql_refs(sql) == ["departments", "employees"]
```
Comment (Light2Dark) [inline tests/_ast/test_sql_visitor.py:604]: i need to investigate more on this ordering. Nested tables are referenced first, I'm not sure what impact that has.

File: pyproject.toml [hunk 2]
```diff
@@ -75,16 +75,22 @@ marimo = "marimo._cli.cli:main"
 homepage = "https://github.com/marimo-team/marimo"
 
 [project.optional-dependencies]
-sql = ["duckdb >= 1.0.0", "polars[pyarrow] >= 1.9.0"]
+sql = [
+    "duckdb>=1.1.0",
```
Comment (akshayka) [inline pyproject.toml:79]: nit: If `duckdb 1.0` is still compatible with this change, we should prefer it over `1.1` so not to break users who haven't upgraded.

File: marimo/_ast/sql_visitor.py [hunk 3]
```diff
@@ -315,109 +313,52 @@ def find_sql_refs(
...
+            # We skip schema if there is a catalog
+            # Because it may be called "public" or "main" across all catalogs
+            # and they aren't referenced in the code
+            if table.catalog:
+                refs.append(table.catalog)
+            elif table.db:
+                refs.append(table.db)  # schema
+
+            if table.name:
+                refs.append(table.name)
```
Comment (akshayka) [inline marimo/_ast/sql_visitor.py:340]: For `my_catalog.my_schema.my_table`, where `my_catalog` is explicitly given in the code, I wonder if the ref should just be `my_catalog` instead of `my_catalog` and `my_table`? Because two separate catalogs could have overlapping table names. Analogous to how in Python, the ref for `my_instance.my_variable` is just `my_instance`. 

A genuine question. Curious what makes more sense to you.

**Review by akshayka — COMMENTED** [review]
This is fantastic, thank you so much! I'm very excited by what a migration to sqlglot will unlock in the future, including support for other dialects.

Left one nit and one question.

File: marimo/_ast/sql_visitor.py (hunk 3 above)
Comment (akshayka) [reply marimo/_ast/sql_visitor.py:340]: I think right now, `my_catalog.my_schema.my_table` generates a def of just `my_table` -- so the logic of adding `my_table` as a ref makes sense in the context of this PR. Perhaps we can revisit this question later.

Comment (mscolnick) [reply marimo/_ast/sql_visitor.py:340]: Yea this was a known limitation / edge case when we added this, but it was the lesser of two evils. If using sqlglot always gives us a valid catalog, we could only use fully qualified/verbose table names. 

But given this does the same logic as before, I think we can look into it later (agreed with @akshayka)

Comment (Light2Dark) [reply marimo/_ast/sql_visitor.py:340]: can revisit this when we modify `find_sql_defs` too

File: marimo/_ast/sql_visitor.py [hunk 4]
```diff
@@ -315,109 +313,52 @@ def find_sql_refs(
...
+
+    expression_list = parse(sql_statement)
```
Comment (akshayka) [inline marimo/_ast/sql_visitor.py:342]: Since we only support `duckdb` right now, and since duckdb has some special functions/syntax that other dialects don't have, I think this should be:

```suggestion
    expression_list = parse(sql_statement, dialect="duckdb")
```

**Review by akshayka — COMMENTED** [review]
Just one comment, and then looks good to me. Thanks for this change!

File: tests/_ast/test_sql_visitor.py [hunk 5]
```diff
@@ -626,18 +626,6 @@ def test_find_sql_refs_nested_intersect() -> None:
...
-    def test_find_sql_refs_with_recursive_cte() -> None:
-        sql = """
-        WITH RECURSIVE cte AS (
-            SELECT 1 AS n FROM table1
-            UNION ALL
-            SELECT n + 1 FROM cte WHERE n < 10
-        )
-        SELECT * FROM cte;
-        """
-        assert find_sql_refs(sql) == ["table1"]
```
Comment (akshayka) [inline tests/_ast/test_sql_visitor.py]: If this is not yet supported, but you'd like to keep the test, you can decorate it with `pytest.mark.xfail`, as done here: https://github.com/marimo-team/marimo/blob/f8dc9f5c210a44375f5a431771efcfc9a6f17ea7/tests/_ast/test_sql_visitor.py#L569

Comment (Light2Dark) [reply tests/_ast/test_sql_visitor.py]: it does work, I didn't see the need for it since we have tested nested cte's. Also noted on the dialect change. I may add back the has_duckdb dependency too.

File: tests/_ast/test_sql_visitor.py [hunk 6]
```diff
@@ -726,3 +726,8 @@ def test_find_sql_refs_select_in_delete() -> None:
         );
         """
         assert find_sql_refs(sql) == ["table1", "table2"]
+
+    @staticmethod
+    def test_find_sql_refs_invalid_sql() -> None:
+        sql = "SELECT * FROM"
+        assert find_sql_refs(sql) == []
```
Comment (akshayka) [inline tests/_ast/test_sql_visitor.py:733]: Nice test, thanks

**Review by akshayka — APPROVED** [review]
Fantastic work! This will have a large impact on marimo's SQL support; thanks so much for getting the ball rolling!
